### PR TITLE
check for remeshed DEM, save remeshed DEM

### PR DESCRIPTION
**Pull request by awirb**
* if desired meshCellSize in configuration doesn't match approx. cellSize of DEM in Inputs/ - go to DEMremeshed and load one .asc file after the other if a DEM found with good cellSize -> stop and use this one
* if none found -> remesh 
* if remeshing is performed -> save new DEM to Inputs/DEMremeshed as remeshedDEM%.2f.asc % meshCellSize
* close issue #670

### CODE REVIEW

File: avaframe/in3Utils/geoTrans.py [hunk 1]
```diff
@@ -288,35 +291,85 @@ def remeshDEM(cfg, dem):
...
+            flag if dem is found
+    """
+
+    # load input parameters
+    cellSize = cfg.getfloat('meshCellSize')
+    cellSizeThreshold = cfg.getfloat('meshCellSizeThreshold')
+
+    # path to remeshed DEM folder
+    pathToDem = pathlib.Path(cfg['avalancheDir'], 'Inputs', 'DEMremeshed')
+    fU.makeADir(pathToDem)
```
Comment (matthiasto) [inline avaframe/in3Utils/geoTrans.py:360]: Why make the folder? You only need to read it no?

File: avaframe/in3Utils/geoTrans.py [hunk 2]
```diff
@@ -288,35 +291,85 @@ def remeshDEM(cfg, dem):
...
+    demFiles = list(pathToDem.glob('*.asc'))
+    DEMFound = False
+    dem = {}
+    for demF in demFiles:
+        demDict = IOf.readRaster(demF)
+        if abs(cellSize - demDict['header']['cellsize']) < cellSizeThreshold:
+            log.info('Remeshed DEM found: %s cellSize: %.5f' % (demF.name, demDict['header']['cellsize']))
+            DEMFound = True
+            dem = demDict
+            continue
```
Comment (matthiasto) [inline avaframe/in3Utils/geoTrans.py:372]: suggestion. Maybe we could make kind of a csv file with the info about all remeshed dem. Then you could look at this dataframe to choose the good dem?

File: avaframe/in3Utils/geoTrans.py [hunk 3]
```diff
@@ -288,35 +291,85 @@ def remeshDEM(cfg, dem):
...
-        dem['header'] = headerRemeshed
-        xNewGrid, yNewGrid = np.meshgrid(xNew, yNew)
-        zNew = sp.interpolate.griddata((xGrid, yGrid), z, (xNewGrid, yNewGrid), method='cubic',
-                                       fill_value=headerDEM['noDataValue'])
-        log.info('Remeshed data extent difference x: %f and y %f' % (diffExtentX, diffExtentY))
-        dem['rasterData'] = zNew
-
-    return dem
+        # first check if remeshed DEM is available
+        remeshedDem, DEMFound = searchRemeshedDEM(cfg)
```
Comment (matthiasto) [inline avaframe/in3Utils/geoTrans.py:295]: maybe add an option to force the remeshing even if the re meshed dem already exists? And make it the default one.

**Review by matthiasto — CHANGES_REQUESTED** [review]

File: avaframe/in3Utils/geoTrans.py (hunk 2 above)
Comment (awirb) [reply avaframe/in3Utils/geoTrans.py:372]: how should we do this? every time a remeshedDEM is stored we open the csv and append the info? but then it is not allowed to store a DEM from outside there, and if we read the DEM first to create the csv. it takes more time

Comment (fso42) [reply avaframe/in3Utils/geoTrans.py:372]: I´m not a fan of creating a csv! We are not talking about a common occurance with tons of files. Reading a dem and getting the header is not huge expense...

File: avaframe/in3Utils/geoTrans.py (hunk 1 above)
Comment (awirb) [reply avaframe/in3Utils/geoTrans.py:360]: this directory only exists if before a remeshed DEM has been saved, hence when searching - I changed it to check if dir exists

File: avaframe/in3Utils/geoTrans.py (hunk 3 above)
Comment (awirb) [reply avaframe/in3Utils/geoTrans.py:295]: I added a flag forceRemesh in com1DFACfg.ini - if True force remeshing - now default

File: avaframe/com1DFA/com1DFA.py [hunk 4]
```diff
@@ -267,8 +267,7 @@ def com1DFACore(cfg, avaDir, cuSimName, inputSimFiles, outDir):
     # +++++++++PERFORM SIMULAITON++++++++++++++++++++++
     # for timing the sims
     startTime = time.time()
-    particles, fields, dem, reportAreaInfo = initializeSimulation(cfg, demOri, inputSimLines, cuSimName)
-
+    particles, fields, dem, reportAreaInfo, demOri = initializeSimulation(cfg, demOri, inputSimLines, cuSimName)
```
Comment (fso42) [inline avaframe/com1DFA/com1DFA.py:270]: I think the logic for demOri vs dem needs to be tackled:
- If there is demOri as input, but an additional dem as output, I think the basic (and valid) assumption is, that demOri remains unchanged and `dem` contains the topography to use -> then there's no need to return demOri
- If demOri is somehow changed, why then the need for an additional `dem` variable?

I suggest to make it a clear distinction: demOri -> the originial, untouched DEM, dem -> the applied, and maybe modified, dem

File: avaframe/com1DFA/com1DFACfg.ini [hunk 5]
```diff
@@ -198,11 +198,13 @@ cMaxnMass = 5
...
 # minimum flow depth [m]
 hmin = 0.05
-# remesh the input DEM
+# remesh the input DEM or look for remeshed DEM 
 # expected mesh size [m]
 meshCellSize = 5
 # threshold under which no remeshing is done
 meshCellSizeThreshold = 0.001
+# force remesh - if True remeshing performed also if remeshed DEMs in Inputs/DEMremeshed available
+forceRemesh = True
```
Comment (fso42) [inline avaframe/com1DFA/com1DFACfg.ini:207]: why is a forceRemesh necessary? Either the mesh is different from the specified resolution -> remesh, or it is the same -> do not remesh...

File: avaframe/in3Utils/geoTrans.py [hunk 6]
```diff
@@ -286,8 +289,16 @@ def remeshDEM(cfg, dem):
...
-    # remesh if input DEM size does not correspond to the computational cellSize
-    if np.abs(cszNew - cszDEM) > cszThreshold:
+    remesh = True
+    if cfg.getboolean('forceRemesh') is False:
+        # first check if remeshed DEM is available
+        remeshedDEM, DEMFound = searchRemeshedDEM(cfg, dem)
+        if DEMFound:
+            remesh = False
+    else:
+        log.info('Forcing remeshing')
```
Comment (fso42) [inline avaframe/in3Utils/geoTrans.py:299]: see comment about forceRemesh

File: avaframe/in3Utils/geoTrans.py [hunk 7]
```diff
@@ -286,8 +289,16 @@ def remeshDEM(cfg, dem):
...
+
+    if remesh:
```
Comment (fso42) [inline avaframe/in3Utils/geoTrans.py:301]: drop this  if and add a return above

File: avaframe/in3Utils/geoTrans.py [hunk 8]
```diff
@@ -309,14 +320,68 @@ def remeshDEM(cfg, dem):
...
+    cellSize = cfg.getfloat('meshCellSize')
+    cellSizeThreshold = cfg.getfloat('meshCellSizeThreshold')
+
+    # path to remeshed DEM folder
+    pathToDem = pathlib.Path(cfg['avalancheDir'], 'Inputs', 'DEMremeshed')
+    DEMFound = False
+    remeshedDEM = {}
+    if pathToDem.is_dir():
+        # look for dems and check if cellSize within tolerance and origin matches
+        demFiles = list(pathToDem.glob('*.asc'))
```
Comment (fso42) [inline avaframe/in3Utils/geoTrans.py:369]: split this here and move to for to the same indent as the next for. Do a else return/etc on the if pathToDem

File: avaframe/in3Utils/geoTrans.py [hunk 9]
```diff
@@ -309,14 +320,68 @@ def remeshDEM(cfg, dem):
...
+        for demF in demFiles:
+            headerDEM = IOf.readASCheader(demF)
+            if abs(cellSize - headerDEM['cellsize']) < cellSizeThreshold:
+                if ('%.2f' % headerDEM['xllcenter']) == ('%.2f' % dem['header']['xllcenter']) and ('%.2f' % headerDEM['yllcenter']) == ('%.2f' % dem['header']['yllcenter']):
+                    remeshedDEM = IOf.readRaster(demF)
+                    log.info('Remeshed DEM found: %s cellSize: %.5f' % (demF.name, remeshedDEM['header']['cellsize']))
+                    DEMFound = True
+                    continue
+                else:
+                    log.warning('Remeshed dem found with cellSize %.2f but origin not matching' % cellSize)
```
Comment (fso42) [inline avaframe/in3Utils/geoTrans.py:379]: Is checking cellsize and origin really enough? Should we check the name as well? Is the origin even needed?

**Review by fso42 — CHANGES_REQUESTED** [review]

**Comment by fso42** [comment]
Maybe consider #658  as well?

**Comment by awirb** [comment]
see new PR #677

**Pull request closed without merging**